TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java

```diff
@@ -182,11 +191,44 @@ private void writeTimestampForNowToCurrentStream(Instant timestamp) throws IOExc
         }
     }
 
-    private void writeByteBufferToCurrentStream(int fieldNum, ByteBuffer byteBuffer) throws IOException {
-        if (byteBuffer.remaining() > 0) {
-            getOrCreateCodedOutputStream().writeByteBuffer(fieldNum, byteBuffer);
+    // Optimistically calculates maxWriteableSpace taking into account lengthSpace. In some cases, this may
+    // yield a maxWriteableSpace with one leftover byte, however, this is still the correct max as attempting to write
+    // the additional byte would yield an additional length byte and overflow
+    public static int calculateMaxWritableSpace(int totalAvailableSpace, int requestedWriteableSpace) {
+        // Overestimate the lengthFieldSpace first to then correct in instances where writing one more byte does not result
+        // in as large a lengthFieldSpace. In instances where we must have a leftoverByte, it will be in the lengthFieldSpace
+        final int maxLengthFieldSpace = CodedOutputStream.computeUInt32SizeNoTag(totalAvailableSpace);
+        final int initialEstimatedMaxWriteSpace = totalAvailableSpace - maxLengthFieldSpace;
+        final int lengthFieldSpaceForInitialEstimatedAndOneMoreByte = CodedOutputStream.computeUInt32SizeNoTag(initialEstimatedMaxWriteSpace + 1);
+
+        int maxWriteBytesSpace = totalAvailableSpace - lengthFieldSpaceForInitialEstimatedAndOneMoreByte;
+
+        return Math.min(maxWriteBytesSpace, requestedWriteableSpace);
+    }
+
+    // Similar to calculateMaxWritableSpace but perform pessimistic calculation with fewer operations. In some cases returns
+    // up to 1 byte fewer than what could be written out of the available space.
+    public static int pessimisticallyCalculateMaxWritableSpace(int totalAvailableSpace, int requestedWriteableSpace) {
```

Comment:
@gregschohn, added, depended on, and tested pessimisticallyCalculateMaxWritableSpace instead of calculateMaxWritableSpace

Reply:
I'm thinking `computeMaxLengthDelimitedFieldSizeForSpace` which is analogous to (internal method) `computeLengthDelimitedFieldSize` in CodedOutputStream